core/src/main/scala/org/apache/spark/serializer/JavaSerializer.scala

```diff
@@ -58,19 +59,32 @@ private[spark] class JavaSerializationStream(
   def close() { objOut.close() }
 }
 
-private[spark] class JavaDeserializationStream(in: InputStream, loader: ClassLoader)
-  extends DeserializationStream {
+private[spark] class JavaDeserializationStream(
+    in: InputStream,
+    loader: ClassLoader,
+    useCache: Boolean,
+    serializerInstance: JavaSerializerInstance) extends DeserializationStream {
 
   private val objIn = new ObjectInputStream(in) {
-    override def resolveClass(desc: ObjectStreamClass): Class[_] =
+    override def resolveClass(desc: ObjectStreamClass): Class[_] = {
+      if (useCache) {
+        serializerInstance.resolvedClassesCache.computeIfAbsent(
+          (desc.getName, loader), pair => normalResolve(pair._1))
+      } else {
+        normalResolve(desc.getName)
+      }
+    }
+
+    private def normalResolve(name: String): Class[_] = {
       try {
         // scalastyle:off classforname
-        Class.forName(desc.getName, false, loader)
+        Class.forName(name, false, loader)
```

Comment:
@rednaxelafx does JVM classloader have cache?

Reply:
Yeah, same question. It's native code so hard to see. I suspect it does, but we may still be saving some overhead in digging it out of the JVM?

Reply:
Class.forName seems is not free. I have done a simple test in local.
```scala
object ResolveClass {
  def main(args: Array[String]): Unit = {
    var i = 0
    while (i < 10) {
      Class.forName(Person.getClass.getName)
      i += 1
    }
    val start = System.currentTimeMillis()
    i = 0
    while (i < 100000) {
      Class.forName(Person.getClass.getName)
      i += 1
    }

    println(System.currentTimeMillis() - start)
  }
}
```
it needs about 196 ms.